### ISSUE TITLE
Inconsistent pegs4 density correction file paths

### ISSUE THREAD

**Issue by ftessier** [nrc-cnrc/EGSnrc#67]
## Issue

The paths for user-supplied pegs4 density correction files is inconsistent across EGSnrc:

1. `HEN_HOUSE/scripts/configure` creates a `egs_home/pegs4/density` directory
2. `get_media_inputs.mortran` looks in the `egs_home/pegs4/density` created during configuration (since commit b53023b)
3. but `pegs4.mortran` (and thus `egs_gui`) looks for density correction files inside `egs_home` paths similar to those found under `HEN_HOUSE`, i.e., `egs_home/pegs4/density_corrections/compounds` and `egs_home/pegs4/density_corrections/elements`

What is most confusing is that `egs_gui` allows the user to load a `.density` file from anywhere, but yields an empty `.pegs4dat` file if the density file is not found in the canonical paths `egs_home/pegs4/density_corrections/compounds` or `egs_home/pegs4/density_corrections/elements`. Regardless, `egs_gui` reports that "PEGS finished succesfully".

## Proposed solution

1. `configure` creates a `egs_home/pegs4/density_corrections` directory (`density` is a little vague, and different from the `HEN_HOUSE` name)
2. `get_media_inputs.mortran` searches `egs_home/pegs4/density_corrections`
3. `pegs4.mortran` searches `egs_home/pegs4/density_corrections` (without the `elements` or `compounds` sub-directories)
4. `egs_gui` restricts loading files to the system's `HEN_HOUSE/pegs4/density_corrections/` (`elements`, `compounds`) directories and the user's `egs_home/pegs4/density_corrections` directory

#### egs_gui success but empty pegs4 data file

1. Create (or copy) a `.density` file in `egs_home/pegs4/density`
2. Start `egs_gui` from home directory: `(cd; egs_gui &)`; go to PEGS Data page
3. Load the `.density` file from step 1, specify energies, material name, file name and click "Go"
4. The `egs_gui` reports that "PEGS finished succesfully"
6. The pegs file is generated in `egs_home/pegs4/data`, but it has 0 bytes.
5. In fact the "Details>>" window in `egs_gui` does not report using any density correction file and further reveals that 

        At line 1978 of file pegs4_linux.f (unit = 20, file = 'fort.20')
        Fortran runtime error: End of file

#### egs_gui success if launched from .density file directory

1. Create (or copy) a `.density` file your home directory `~/`
2. Start `egs_gui` from home directory: `(cd; egs_gui &)`; go to PEGS Data page
3. Load the `.density` file from step 1, specify energies, material name, file name and click "Go"
4. The `egs_gui` reports that "PEGS finished succesfully"
5. The pegs file is generated in `egs_home/pegs4/data`, which seems fine
6. However, the "Details>>" window in `egs_gui` **does NOT** report the path of the density file:

        ************ Using density file fred.density

        ************ Using cross section data file 
        /home/user/EGSnrc/HEN_HOUSE/pegs4/pgs4pepr.dat

        ************ Using form factor data file
        /home/user/EGSnrc/HEN_HOUSE/pegs4/pgs4form.dat

#### egs_gui success with `density_corrections` directories

1. `mkdir -p $EGS_HOME/pegs4/density_corrections/{elements,compounds}`
2. Create (or copy) a `.density` file to either `elements` or `compounds` directories created in 1
3. Start `egs_gui` from home directory: `(cd; egs_gui &)`; go to PEGS Data page
4. Load the `.density` file from step 2, specify energies, material name, file name and click "Go"
4. The `egs_gui` reports that "PEGS finished succesfully"
5. The pegs file is generated in `egs_home/pegs4/data`, which seems fine
6. The "Details>>" window in `egs_gui` **does** report the path of the density file:

        ************ Using density file
        /home/user/EGSnrc/egs_home/pegs4/density_corrections/elements/test.density

        ************ Using cross section data file
        /home/user/EGSnrc/HEN_HOUSE/pegs4/pgs4pepr.dat

        ************ Using form factor data file 
        /home/user/EGSnrc/HEN_HOUSE/pegs4/pgs4form.dat

_(Issue initially reported by Brigitte Reniers)_

**Comment by blakewalters** [comment]
I see the same behaviour and clearly this is a bug.  As far as changing the names of the subdirectories off $EGS_HOME/pegs4 in which density correction files are stored to mirror the $HEN_HOUSE/pegs4 structure, I'm not sure that's necessary: Just thinking about how users may elect to store their own custom .density files, it's more convenient to just collect them all in a single directory rather than worry about separating them into elements and compounds subdirectories.  Any thoughts, Fred?

**Comment by ftessier** [comment]
Yes, absolutely, this is what I suggested above:

3. pegs4.mortran searches `egs_home/pegs4/density_corrections` (**without** the `elements` or `compounds` sub-directories)

**Comment by blakewalters** [comment]
How about backward compatibility (i.e. users who are used to pegs4.mortran finding their .density files in $EGS_HOME/pegs4/density_corrections/elements and $EGS_HOME/pegs4/density_corrections/compounds)?  pegs4.mortran has been around for quite awhile now.

**Comment by mainegra** [comment]
I always keep my density corrections files in `$HEN_HOUSE/pegs4/density_corrections`. But of course in a centralized environment this could be tricky. And updates to a $HEN_HOUSE could delete files created by the user.

**Comment by ftessier** [comment]
Fair enough. However as far as I can tell the configuration script never created those directories, users would have have to create them by hand; hence this bug would officially be backwards compatible (!) Perhaps the best solution is to look for the `compounds` and `elements` sub-directories _as well_? In fact, ideally we would recursively find **all** `.density` files under `density_corrections`, but I bet this is not straightforward in mortran.

**Comment by ftessier** [comment]
We should strive not to modify anything manually under HEN_HOUSE, although for the time being to avoid the current issue putting your density correction files there is the best workaround. The other temporary solution under the current release is to change the name of `egs_home/density` to `egs_home/density_corrections` and create the `compounds` and `elements` sub-directories. But that **breaks** pegsless mode, because `get_media_inputs.mortran` looks in `egs_home/pegs4/density` !!

**Comment by blakewalters** [comment]
Here's where pegs4.mortran searches for the .density file (in this order):
1. current directory (presumably for the case where the entire path to the file is specified)
2. $HEN_HOUSE/pegs4/density_corrections/elements
3. $HEN_HOUSE/pegs4/density_corrections/compounds
4. $EGS_HOME/pegs4/density_corrections
5. $EGS_HOME/pegs4/density_corrections/elements
6. $EGS_HOME/pegs4/density_corrections/compounds
I suggest we keep the scope of this search, but reorder it so that $EGS_HOME/pegs4/... is searched after current directory.  In get_media_inputs.mortran, perform this same search but also include the density subdirectory (to keep this fix backwards compatible).  As for egs_gui, I suggest we continue with browsing in $HEN_HOUSE/pegs4/density_corrections or $EGS_HOME/pegs4/density_corrections.  If the user has  $EGS_HOME/pegs4/density defined instead, the browser automatically starts in $EGS_HOME/pegs4, from which the user can then navigate into the correct subdirectory and select the file.  Provided egs_gui supplies pegs4.mortran with the full path name to the density file (currently not done), all should be well.

**Comment by ftessier** [comment]
Agreed with your proposed search order and backward compatibility.

I had missed the check for the current directory, which explains why `egs_gui` succeeded when launching from the directory where the `.density` file is stored. While this is arguably useful, in my case it was confusing because sometimes for the same file it worked, and sometimes not, depending on where I launched `egs_gui` from!

At any rate go ahead and update `pegs4.mortran`, and `get_media_inputs.mortran` on a branch `fix-67-density-paths` or the like, and push it to github:
```
git checkout develop
git pull
git branch -d fix-67-density-paths
(make changes)
git commit
git push origin fix-67-density-paths
```
We'll then kindly ask @mainegra to adjust `egs_gui` on that branch, accordingly.

**Comment by blakewalters** [comment]
Alright.  If @mainegra's okay with it I can fix up egs_gui as well!

**Comment by mainegra** [comment]
Go for it @blakewalters!

**Comment by ftessier** [comment]
**git branch -b** in my previous post!!! Create a new branch, not delete it!!

**Comment by ftessier** [comment]
Fixed in 1de38ba
